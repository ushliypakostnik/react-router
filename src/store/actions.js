const fetchUrl = process.env.REACT_APP_API_URL;

export const pageToActive = (page) => ({
  type: "PAGE_TO_ACTIVE",
  activePage: page
});

export const setMinHeight = (minHeight) => ({
  type: "SET_MIN_HEIGHT",
  minHeight: minHeight
});

export const toogleTheme = (theme) => ({
  type: "TOGGLE_THEME",
  theme: theme
});

export const requestAlbums = () => ({
  type: "REQUEST_ALBUMS"
});

export const receiveAlbums = (albums) => ({
  type: "RECEIVE_ALBUMS",
  albums
});

export const requestAlbumFailed = (error) => ({
  type: "REQUEST_ALBUMS_FAILED",
  error
});

export const fetchAlbums = () => {
  return dispatch => {
    dispatch(requestAlbums());
    return fetch(fetchUrl + "/albums")
      .then(res => res.json())
      .then(
        (result) => {
          /* eslint-disable array-callback-return */
          let resultArr = [];
          result.map((item, index) => {
            if (index === 0) {
              resultArr.push({text: item.name, path: "/"});
            } else {
              resultArr.push({text: item.name, path: "/" + item.id});
            }
          });
          dispatch(receiveAlbums(resultArr));
        },
        (error) => {
          dispatch(requestAlbumFailed(error));
        }
      );
  }
}

export const requestData = () => ({
  type: "REQUEST_DATA"
});

export const receiveData = (data) => ({
  type: "RECEIVE_DATA",
  data
});

export const requestDataFailed = (error) => ({
  type: "REQUEST_DATA_FAILED",
  error
});

export const fetchData = (album) => {
  return dispatch => {
    dispatch(requestData());
    return fetch(fetchUrl + "/albums/" + album)
      .then(res => res.json())
      .then(
        (result) => {
          dispatch(receiveData(result));
        },
        (error) => {
          dispatch(requestDataFailed(error));
        }
      );
  }
}
