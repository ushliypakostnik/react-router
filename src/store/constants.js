export const FETCH_URL = process.env.REACT_APP_API_URL;

export const THEME = {
  DARK: "dark",
  LIGHT: "light"
};

export const COOKIES = {
  THEME: "theme"
};
