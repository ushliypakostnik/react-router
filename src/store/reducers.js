const reducer = (state = { }, action) => {
  switch (action.type) {
    case "REQUEST_ALBUMS":
      return Object.assign({}, state, {
        isFetching: true
      });
    case "RECEIVE_ALBUMS":
      return Object.assign({}, state, {
        isFetching: false,
        albums: action.albums
      });
    case "REQUEST_ALBUMS_FAILED":
      return Object.assign({}, state, {
        isFetching: false,
        error: action.error
      });
    case "REQUEST_DATA":
      return Object.assign({}, state, {
        isFetching: true
      });
    case "RECEIVE_DATA":
      return Object.assign({}, state, {
        isFetching: false,
        data: action.data
      });
    case "REQUEST_DATA_FAILED":
      return Object.assign({}, state, {
        isFetching: false,
        error: action.error
      });
    case "PAGE_TO_ACTIVE":
      return Object.assign({}, state, {
        activePage: action.activePage
      });
    case "TOGGLE_THEME":
      return Object.assign({}, state, {
        theme: action.theme
      });
    default:
      return state;
  }
}

export default reducer;
