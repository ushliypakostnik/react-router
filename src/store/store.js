import { createStore, combineReducers, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import { createLogger } from 'redux-logger';
import { browserHistory } from "react-router";
//import { createBrowserHistory } from "history";
import { syncHistoryWithStore, routerReducer } from 'react-router-redux';

import { INITIAL_STATE } from './constants';
import { pageToActive } from './actions';
import reducer from './reducers';

const middlewares = [];
middlewares.push(thunkMiddleware)

if (process.env.NODE_ENV !== 'production') {
  const loggerMiddleware = createLogger();

  middlewares.push(loggerMiddleware);
}

function configureStore(state) {
  return createStore(
    combineReducers({
      reducer,
      routing: routerReducer
    }),
    state,
    applyMiddleware(...middlewares)
  );
}

const store = configureStore(INITIAL_STATE);

//const history = createBrowserHistory();
export const history = syncHistoryWithStore(browserHistory, store);
//history.listen(location => console.info('-> location:', location));

let lastUrl = "";
if (store.getState().routing.locationBeforeTransitions != null) {
  lastUrl = store.getState().routing.locationBeforeTransitions.pathname;
} else {
  lastUrl = "/";
}

//store.subscribe(() => {
//  console.log("Store: ", store.getState());
//  console.log("last Url: ", lastUrl);
//});

store.dispatch(pageToActive(lastUrl));

export default store;
