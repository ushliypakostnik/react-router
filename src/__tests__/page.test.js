import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { FETCH_URL, INITIAL_STATE } from '../store/constants';

import Page from '../containers/page';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('page render', () => {

  const props = {
    path: '/pinhole',
    minHeight: '700px',
    isLoaded: false,
    error: null,
    images: []
  };

  const state = Object.assign({}, INITIAL_STATE, {
    reducer: {
      ...INITIAL_STATE.reducer,
      isFetching: !props.isLoaded,
      resize: { minHeight: props.minHeight }
    }
  });
  const store = mockStore(state);

  const wrapper = mount(<Provider store={store}>
                          <Page {...props} />
                        </Provider>);

  it('page render correctly: initial state', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('page contain wrapper with class', () => {
    expect(wrapper.find('.app__page')).toHaveLength(1);
  });

  it('page componentDidMount', () => {
    const spy = jest.spyOn(wrapper.find('Page').instance(), 'componentDidMount');
    wrapper.find('Page').instance().componentDidMount();
    expect(spy).toHaveBeenCalled();
  });

  describe('page render: data is loaded', () => {

    const props1 = {
      ...props,
      isLoaded: true,
      images: [
        { src: `${FETCH_URL}/media/pinhole/010.jpg`, width: 620, height: 632 },
        { src: `${FETCH_URL}/media/pinhole/020.jpg`, width: 1568, height: 632 },
        { src: `${FETCH_URL}/media/pinhole/030.jpg`, width: 1053, height: 632 }
      ]
    };

    const state1 = Object.assign({}, INITIAL_STATE, {
      reducer: {
        ...INITIAL_STATE.reducer,
        isFetching: !props1.isLoaded,
        data: props1.images,
        resize: { minHeight: props1.minHeight }
      }
    });
    const store1 = mockStore(state1);

    const wrapper1 = mount(<Provider store={store1}>
                            <Page {...props1} />
                          </Provider>);

    it('page render correctly: data is loaded', () => {
      expect(toJson(wrapper1)).toMatchSnapshot();
    });

    it('page contain album', () => {
      expect(wrapper.find('Album')).toBeTruthy();
    });
  });

  describe('page render: data error', () => {

    const props2 = {
      ...props,
      error: 'error'
    };

    const state2 = Object.assign({}, INITIAL_STATE, {
      reducer: {
        ...INITIAL_STATE.reducer,
        error: props2.error,
        isFetching: !props2.isLoaded,
        resize: { minHeight: props2.minHeight }
      }
    });
    const store2 = mockStore(state2);

    const wrapper2 = mount(<Provider store={store2}>
                            <Page {...props2} />
                          </Provider>);

    it('page render correctly: data error', () => {
      expect(toJson(wrapper2)).toMatchSnapshot();
    });
  });
});
