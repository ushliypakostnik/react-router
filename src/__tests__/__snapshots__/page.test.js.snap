// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page render page render correctly: initial state 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Page)
    error={null}
    images={Array []}
    isLoaded={false}
    minHeight="700px"
    path="/pinhole"
  >
    <Page
      error={null}
      fetchData={[Function]}
      images={Array []}
      isLoaded={false}
      minHeight="700px"
      path="/pinhole"
    >
      <div
        className="app__page"
        style={
          Object {
            "minHeight": "700px",
          }
        }
      />
    </Page>
  </Connect(Page)>
</Provider>
`;

exports[`page render page render: data error page render correctly: data error 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Page)
    error="error"
    images={Array []}
    isLoaded={false}
    minHeight="700px"
    path="/pinhole"
  >
    <Page
      error="error"
      fetchData={[Function]}
      images={Array []}
      isLoaded={false}
      minHeight="700px"
      path="/pinhole"
    >
      <div
        className="app__page"
        style={
          Object {
            "minHeight": "700px",
          }
        }
      >
        Error: 
      </div>
    </Page>
  </Connect(Page)>
</Provider>
`;

exports[`page render page render: data is loaded page render correctly: data is loaded 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Page)
    error={null}
    images={
      Array [
        Object {
          "height": 632,
          "src": "http://127.0.0.1:8082/media/pinhole/010.jpg",
          "width": 620,
        },
        Object {
          "height": 632,
          "src": "http://127.0.0.1:8082/media/pinhole/020.jpg",
          "width": 1568,
        },
        Object {
          "height": 632,
          "src": "http://127.0.0.1:8082/media/pinhole/030.jpg",
          "width": 1053,
        },
      ]
    }
    isLoaded={true}
    minHeight="700px"
    path="/pinhole"
  >
    <Page
      error={null}
      fetchData={[Function]}
      images={
        Array [
          Object {
            "height": 632,
            "src": "http://127.0.0.1:8082/media/pinhole/010.jpg",
            "width": 620,
          },
          Object {
            "height": 632,
            "src": "http://127.0.0.1:8082/media/pinhole/020.jpg",
            "width": 1568,
          },
          Object {
            "height": 632,
            "src": "http://127.0.0.1:8082/media/pinhole/030.jpg",
            "width": 1053,
          },
        ]
      }
      isLoaded={true}
      minHeight="700px"
      path="/pinhole"
    >
      <section
        className="app__page"
        style={
          Object {
            "minHeight": "700px",
          }
        }
      >
        <Connect(Album)
          photos={
            Array [
              Object {
                "height": 632,
                "src": "http://127.0.0.1:8082/media/pinhole/010.jpg",
                "width": 620,
              },
              Object {
                "height": 632,
                "src": "http://127.0.0.1:8082/media/pinhole/020.jpg",
                "width": 1568,
              },
              Object {
                "height": 632,
                "src": "http://127.0.0.1:8082/media/pinhole/030.jpg",
                "width": 1053,
              },
            ]
          }
        >
          <Album
            dispatch={[Function]}
            photos={
              Array [
                Object {
                  "height": 632,
                  "src": "http://127.0.0.1:8082/media/pinhole/010.jpg",
                  "width": 620,
                },
                Object {
                  "height": 632,
                  "src": "http://127.0.0.1:8082/media/pinhole/020.jpg",
                  "width": 1568,
                },
                Object {
                  "height": 632,
                  "src": "http://127.0.0.1:8082/media/pinhole/030.jpg",
                  "width": 1053,
                },
              ]
            }
          >
            <div
              className="app__gallery"
            >
              <Gallery
                direction="row"
                margin={0}
                onClick={[Function]}
                photos={
                  Array [
                    Object {
                      "height": 632,
                      "src": "http://127.0.0.1:8082/media/pinhole/010.jpg",
                      "width": 620,
                    },
                    Object {
                      "height": 632,
                      "src": "http://127.0.0.1:8082/media/pinhole/020.jpg",
                      "width": 1568,
                    },
                    Object {
                      "height": 632,
                      "src": "http://127.0.0.1:8082/media/pinhole/030.jpg",
                      "width": 1053,
                    },
                  ]
                }
                targetRowHeight={350}
              >
                <div>
                  Â 
                </div>
              </Gallery>
            </div>
          </Album>
        </Connect(Album)>
      </section>
    </Page>
  </Connect(Page)>
</Provider>
`;
